Implement-Permissions-for-Investors-(issue-#70)

DimaMakovskyi opened the pull request:
feat: implement subscription creation flow with proper error handling

- Return 403 when user is not authenticated or not an investor
- On valid subscription by investo
- Added tests for unauthorized, non-investor, and successful investor cases

MaksymZymyn1989 reviewed (approved):
This is a well-scoped and high-quality PR that clearly addresses permissioning for investment actions. Thank you.

DimaMakovskyi commented:
Changed the search tests:

Full data for models in SetUp

To avoid IntegrityError, add all required fields (category, email, current_financing, etc.) for startup and project.

Now objects are not filtered due to the presence of the required data.

Authorization

Added self.client.force_authenticate(user=self.user) to avoid 403 and test with an already authorized user.

Elasticsearch mocks

Moked the Search object that is used in views (search.views), not the global one.

Made the search iterable: so that it works without .execute() to get into the search.

Added support for slices (__getitem__) and method chains (query, filter, sort).

DimaMakovskyi commented on tests/investments/test_subscription_permissions.py:
```diff
@@ -0,0 +1,130 @@
+
+        # Verify that the project was updated
+        self.project.refresh_from_db()
+        self.assertEqual(self.project.current_funding, Decimal("100.00"))
```
fix

DimaMakovskyi commented on users/permissions.py:
```diff
@@ -115,3 +115,39 @@ def has_permission(self, request, view):
+            user.id,
+            view.__class__.__name__,
+        )
+        return True
```
fix

DimaMakovskyi commented on users/permissions.py:
```diff
@@ -115,3 +115,39 @@ def has_permission(self, request, view):
+            )
+            return False
+
+        is_investor = Investor.objects.filter(user=user).exists()
```
fix

DimaMakovskyi commented on investments/views.py:
```diff
@@ -1,79 +1,115 @@
-        )
\ No newline at end of file
+            project=project,
+        )
```
fix

DimaMakovskyi commented on investments/views.py:
```diff
@@ -1,79 +1,115 @@
-                status=status.HTTP_201_CREATED,
-            )
+        try:
+            self.perform_create(serializer)
```
fix

DimaMakovskyi commented on investments/views.py:
```diff
@@ -1,79 +1,115 @@
+                "subscription": SubscriptionCreateSerializer(
+                    serializer.instance, context={"request": request}
+                ).data,
+                "remaining_funding": remaining,  # Decimal, not string
```
fix

mehalyna commented on tests/search/test_search.py:
```diff
@@ -1,106 +1,140 @@
+
+        search_obj.__getitem__.return_value = search_obj
+
+        search_obj.__iter__ = lambda s: iter(hits)
```
Overriding iter with a lambda on a mock may not work with all mock usages. Use mock.Mock's spec and __iter__ setup for better compatibility.
```
search_obj.__iter__.side_effect = lambda: iter(hits)
```

mehalyna commented on tests/search/test_search.py:
```diff
@@ -1,106 +1,140 @@
+        for m in ("query", "filter", "source", "extra"):
+            getattr(search_obj, m).return_value = search_obj
+
+        search_obj.__getitem__.return_value = search_obj
```
Mocking getitem directly may not work as expected for slicing. Use side_effect with slice handling for more robust mocks.
```
search_obj.__getitem__.side_effect = lambda s: search_obj if isinstance(s, slice) else hits[s]
```

mehalyna commented on tests/search/test_search.py:
```diff
@@ -1,106 +1,140 @@
+            self._chainable_search(search_obj, [hit], qs=qs)
+
+            url = reverse("startup-search")
+            resp = self.client.get(url, {"q": "health"}) 
```
Hardcoded query "health" may make the test brittle if search logic changes. Use a variable for the search term for consistency.
```
query = "health"
resp = self.client.get(url, {"q": query})
```

mehalyna commented on tests/search/test_search.py:
```diff
@@ -1,106 +1,140 @@
+            role=role_user,
+            is_active=True,
+        )
+        self.client.force_authenticate(user=self.user)
```
Forcing authentication with a user at setup may cause issues if tests want unauthenticated states. Only force authentication in individual test methods as needed.
